cmd/scollector/collectors/snmp_ciscobgp.go

```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return nil, err
+	}
+	remoteASes := make(map[string]string, len(remoteASesRaw))
+	for k, v := range remoteASesRaw {
+		s := strconv.FormatInt(v.(int64), 10)
+		if err != nil {
+			return nil, err
+		}
+		remoteASes[k] = s
+	}
+
+	// Tag: remote_id
+	remoteIdentifiersRaw, err := snmp_ip_tree(host, community, remoteIdentifier)
+	if err != nil {
+		return nil, err
+	}
+	remoteIdentifiers := make(map[string]string, len(remoteIdentifiersRaw))
+	for k, v := range remoteIdentifiersRaw {
+		remoteIdentifiers[k] = snmp_combine_ip_uint8(v.([]uint8))
+	}
+
+	var md opentsdb.MultiDataPoint
+	// Function to harvest all metrics with the tag groups above
+	add := func(bA bgpAdd) error {
+		m, err := snmp_ip_tree(host, community, bA.oid)
+		if err != nil {
+			return err
+		}
+		for k, v := range m {
+			tags := opentsdb.TagSet{
+				"host":      host,
+				"peer":      k,
+				"local_as":  localASes[k],
+				"local_id":  localIdentifiers[k],
+				"remote_as": remoteASes[k],
+				"remote_id": remoteIdentifiers[k],
+			}
+			Add(&md, bA.metric, v, tags, bA.rate, bA.unit, bA.desc)
+		}
+		return nil
+	}
+	// oid, metric, rate, unit, description
+	oids := []bgpAdd{
+		{state, bgpPeerState, metadata.Gauge, metadata.StatusCode, bgpPeerStateDesc},
+		{adminStatus, bgpPeerAdminStatus, metadata.Gauge, metadata.StatusCode, bgpPeerAdminStatusDesc},
+		{inUpdates, bgpPeerInUpdates, metadata.Counter, metadata.Packet, bgpPeerInUpdatesDesc},
+		{outUpdates, bgpPeerOutUpdates, metadata.Counter, metadata.Packet, bgpPeerOutUpdatesDesc},
+		{inTotalMessages, bgpPeerInTotalMessages, metadata.Counter, metadata.Packet, bgpPeerInTotalMessagesDesc},
+		{outTotalMessages, bgpPeerOutTotalMessages, metadata.Counter, metadata.Packet, bgpPeerOutTotalMessagesDesc},
+		{inUpdateElapsedTime, bgpPeerInUpdateElapsedTime, metadata.Gauge, metadata.Second, bgpPeerInUpdateElapsedTimeDesc},
+		{acceptedPrefixes, bgpPeerAcceptedPrefixes, metadata.Gauge, metadata.Count, bgpPeerAcceptedPrefixesDesc},
+		{deniedPrefixes, bgpPeerDeniedPrefixes, metadata.Counter, metadata.Count, bgpPeerDeniedPrefixesDesc},
+		{advertisedPrefixes, bgpPeerAdvertisedPrefixes, metadata.Counter, metadata.Count, bgpPeerAdvertisedPrefixesDesc},
+		{suppressedPrefixes, bgpPeerSuppressedPrefixes, metadata.Counter, metadata.Count, bgpPeerSuppressedPrefixesDesc},
+		{withdrawnPrefixes, bgpPeerWithdrawnPrefixes, metadata.Counter, metadata.Count, bgpPeerWithdrawnPrefixesDesc},
+	}
+	for _, bA := range oids {
+		if err := add(bA); err != nil {
+			return nil, err
+		}
+	}
+	return md, nil
+}
+
+func snmp_ip_tree(host, community, oid string) (map[string]interface{}, error) {
+	rows, err := snmp.Walk(host, community, oid)
+	if err != nil {
+		return nil, err
+	}
+	m := make(map[string]interface{})
+	for rows.Next() {
+		key := ""
+		var a interface{}
+		id, err := rows.Scan(&a)
+		if err != nil {
+			return nil, err
+		}
+		switch t := id.(type) {
+		case []int:
```

Comment:
I think you can declare a as []int instead of interface{} and pass that into Scan. Then you don;t have to do all of the type checking stuff.